lib/geckodriver.js

```diff
@@ -0,0 +1,152 @@
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.geckodriver = SYSTEM_GECKODRIVER;
+      this.proc = new SubProcess(this.geckodriver, args);
+      processIsAlive = true;
+
+      this.proc.on('output', (stdout, stderr) => {
+        if (this.verbose) {
+          for (let line of (stdout || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.debug(`[STDOUT] ${line}`);
+          }
+          for (let line of (stderr || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.error(`[STDERR] ${line}`);
+          }
+        }
+      });
+
+      this.proc.on('exit', (code, signal) => {
+        log.info(code, signal);
+        processIsAlive = false;
+
+        if (this.state !== Geckodriver.STATE_STOPPED &&
+          this.state !== Geckodriver.STATE_STOPPING &&
+          this.state !== Geckodriver.STATE_RESTARTING) {
+          let msg = `Geckodriver exited unexpectedly with code ${code}, ` +
+                  `signal ${signal}`;
+          log.error(msg);
+          this.changeState(Geckodriver.STATE_STOPPED);
+        }
+      });
+
+      log.info(`Spawning geckodriver with: ${this.geckodriver} ` +
+              `${args.join(' ')}. Only geckodriver 0.26.0 supports Android testing`);
+      // start subproc and wait for startDetector
+      await this.proc.start(startDetector);
+      this.changeState(Geckodriver.STATE_ONLINE);
+      await this.startSession();
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+      if (processIsAlive) {
+        await this.proc.stop();
+      }
+    }
+  }
+
+  sessionId () {
+    if (this.state !== Geckodriver.STATE_ONLINE) {
+      return null;
+    }
+    return this.jwproxy.sessionId;
+  }
+
+  async stop () {
+    try {
+      await this.proc.stop();
+      this.changeState(Geckodriver.STATE_STOPPED);
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+    }
+  }
+
+  async getStatus () {
+    return await this.jwproxy.command('/status', 'GET');
+  }
+
+  prepareMozCapabilities () {
+    let mozFirefoxOptions = {
+      log: {
+        level: 'trace'
+      },
+      androidPackage: 'org.mozilla.firefox',
+      androidActivity: 'org.mozilla.gecko.BrowserApp',
+      androidDeviceSerial: 'FA6980300148',
+      androidIntentArguments: [
+        '-d', this.capabilities.firefoxInitialUrl || DEFAULT_INIT_URL
+      ]
+    };
+    if (!_.has(this.capabilities, 'capabilities.alwaysMatch["moz:firefoxOptions"]') === true) {
+      this.capabilities.capabilities.alwaysMatch['moz:firefoxOptions'] = {};
+    }
+    _.extend(this.capabilities.capabilities.alwaysMatch['moz:firefoxOptions'], mozFirefoxOptions);
+    delete this.capabilities.capabilities.firefoxInitialUrl;
+  }
+
+  async startSession () {
```

Comment:
shouldn't we also override deleteSession?

Reply:
@mykola-mokhnach @christian-bromann added deleteSession but im seeing this error. Have you seen this before:

```
dbug WD Proxy Proxying [DELETE /] to [DELETE http://127.0.0.1:29146/session/4396812b-225e-46ad-acbf-b1e6322c809e] with body: {}
dbug Geckodriver [STDOUT] 1578274046035 webdriver::server       DEBUG   -> DELETE /session/4396812b-225e-46ad-acbf-b1e6322c809e {}
dbug Geckodriver [STDOUT] 1578274046042 webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"unsupported operation","message":"Only supported in Firefox","stacktrace":"WebDriverError@chrome://marionette/content/error.js:4:250\nUnsupportedOperationError@chrome://marionette/content/error.js:37:77\nassert.that/<@chrome://marionette/content/assert.js:5:513\nassert.firefox@chrome://marionette/content/assert.js:1:1325\nGeckoDriver.prototype.quit@chrome://marionette/content/driver.js:133:8\ndespatch@chrome://marionette/content/server.js:21:34\nexecute@chrome://marionette/content/server.js:18:170\nonPacket/<@chrome://marionette/content/server.js:17:50\nonPacket@chrome://marionette/content/server.js:17:65\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:23:125\n"}}
info WD Proxy Got an unexpected response with status 500: {"value":{"error":"unsupported operation","message":"Only supported in Firefox","stacktrace":"WebDriverError@chrome://marionette/content/error.js:4:250\nUnsupportedOperationError@chrome://marionette/content/error.js:37:77\nassert.that/<@chrome://marionette/content/assert.js:5:513\nassert.firefox@chrome://marionette/content/assert.js:1:1325\nGeckoDriver.prototype.quit@chrome://marionette/content/driver.js:133:8\ndespatch@chrome://marionette/content/server.js:21:34\nexecute@chrome://marionette/content/server.js:18:170\nonPacket/<@chrome://marionette/content/server.js:17:50\nonPacket@chrome://marionette/content/server.js:17:65\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:23:125\n"}}
dbug W3C Matched W3C error code 'unsupported operation' to UnsupportedOperationError
```
In the meantime I'll continue reading geckodriver source code 😃

Reply:
DELETE endpoint is a part of W3C spec. It would be strange if it is not supported

Reply:
I'm running the same e2e test locally and firefox desktop is working so something seems to be wrong on Firefox for android. I found this thread that mentions this issue.
https://bugzilla.mozilla.org/show_bug.cgi?id=1525126&GoAheadAndLogIn=1

Reply:
then keep the DELETE call, but just ignore the resulting exception and add the comment stating the actual bug

Reply:
This was fixed in the new firefox 68.5.x version